### PR TITLE
define copy_within on slices

### PR DESCRIPTION
**Pull request by oconnor663**
This is a safe wrapper around `ptr::copy`, for regions within a single slice. Previously, safe in-place copying was only available as a side effect of `Vec::drain`.

I've wanted this API a couple times in the past, and I figured I'd just whip up a PR to help discuss it. It's possible something like this exists elsewhere and I just missed it. It might also be a big enough addition to warrant an RFC, I'm not sure.

### CODE REVIEW

**Comment by BurntPizza** [comment]
Seems like it needs `get_unchecked*` to get rid of the bounds checks despite the `assert`s, unless I'm missing something and that's not safe: https://godbolt.org/z/W3m1kv

**Comment by oconnor663** [comment]
@BurntPizza thanks for catching that. I've amended and rebased.

**Comment by oconnor663** [comment]
Safety question not necessarily specific to this PR: Indexing into a slice relies on `ptr::offset`, which says that its UB to use an offset/index larger than `isize::MAX`. As documented by `ptr::offset`, "memory acquired directly from allocators or memory mapped files *may* be too large to handle with this function." Does that mean that memory mapping a file larger than `isize::MAX` is unsound, because safe code can cause UB just by indexing into it?

**Comment by oconnor663** [comment]
I've published this function as a standalone crate in the meantime: https://crates.io/crates/copy_in_place

**Comment by TimNN** [comment]
Ping from triage @Kimundi / @rust-lang/libs: This PR requires your review.

**Comment by SimonSapin** [comment]
> ```rust
> bytes.copy_in_place(1, 8, 4)
> ```

With three `usize` arguments I feel like I’d need to look up the docs every time in order to figure out which argument is which and what this call does. Maybe range syntax could make calls clearer?

```rust
pub fn copy_in_place(&mut self, src: Range<usize>, dest: RangeFrom<usize>) where T: Copy {
   assert!(src.end >= src.start);
   let count = src.end - src.start;
   // …
}
```
```rust
bytes.copy_in_place(1..5, 8..)

**Comment by oconnor663** [comment]
The name: I noticed there was some precedent along the lines of `drop_in_place` and `grow_in_place` in the standard library. Otherwise I don't have any strong feelings.

The API: Agreed that it's confusing to look at. I think the one major advantage that the current approach has going for it is that it's a clone of `ptr::copy(src, dest, count)`. I didn't think about using ranges before, but I wonder if that would lead users to expect that other types of ranges would work. If not, it could be a little frustrating. But if so, it could lead to other types of confusion, for example `slice.copy_in_place(i.., j..)` (the first presumably ending at the end of the slice, and the second then taking the length of the first, which isn't clear just from looking at it). One of the design thoughts I had, which you preserved, was to leave no room in the API to specify regions of different sizes, and so no need to check that particular condition and panic. But if we were willing to relax that, we could say template on any arg types implementing `std::ops::RangeBounds`. Upsides: Very easy to remember how to call. Downsides: Noisier to read if the caller is starting with a count and needs to compute both range ends. Also possibly more overhead if it's being called inside an exceptionally tight loop.

Tracking issue: @alexcrichton how does that work? Would I file one, or would someone on the core team? Before or after this PR lands?

**Comment by oconnor663** [comment]
I guess another option would be to define some kind of arg struct, so that the caller is forced to name each field that it's passing in. That could help readability, but I don't know if there's precedent for it in the standard library. It would also be unnecessarily verbose in the case where the caller already has variables with decent names.

**Comment by SimonSapin** [comment]
> it's a pretty direct clone of `ptr::copy(src, dest, count)`

But `ptr::copy` has different types of arguments (two pointers and an integer), which conveys information about which argument is which. Compare `copy(a.as_ptr(), b.as_mut_ptr(), 4)` v.s. `copy_in_place(1, 8, 4)`.

Regarding a generic parameter with `RangeBounds`, maybe only for `src`?

**Comment by oconnor663** [comment]
I opened a PR on my crate implementation to take a look at what a range argument would look like. It seems viable. The upside is that the argument types a lot more visible. The downside is that a caller with a `count` instead of a `src_end` needs to do one extra math step before calling (though this counts as an upside too if you already have a `src_end`, so maybe a wash). Also there's one more way to panic, if you supply an inverted range.

**Comment by oconnor663** [comment]
@alexcrichton could you say more about how tracking issues work? Should I file one now, or wait until we have a little more consensus on the API here?

**Comment by alexcrichton** [comment]
@oconnor663 oh sure yeah, so for us a tracking issue is created whenever we merge an unstable feature. So to land this PR we'll want a tracking issue which lists out the various design questions and such.

**Comment by oconnor663** [comment]
I've opened a tracking issue, updated the issue number referenced inline in this PR, and rebased.

File: src/libcore/slice/mod.rs [hunk 1]
```diff
@@ -1618,6 +1618,45 @@ impl<T> [T] {
...
+    /// ```
+    #[unstable(feature = "copy_in_place", issue = "54236")]
+    pub fn copy_in_place(&mut self, src: usize, dest: usize, count: usize) where T: Copy {
+        // Avoid computing `src + count` or `dest + count`, which could overflow.
+        assert!(count <= self.len());
+        assert!(src <= self.len() - count);
+        assert!(dest <= self.len() - count);
+        unsafe {
+            let src_ptr = self.get_unchecked(src) as *const T;
+            let dest_ptr = self.get_unchecked_mut(dest) as *mut T;
```
Comment (scottmcm) [inline src/libcore/slice/mod.rs]: nit: You shouldn't need to cast these; they'll coerce just fine.

**Comment by scottmcm** [comment]
> Previously, safe in-place copying was only available as a side effect of Vec::drain.

Would we only suggest this when things can overlap?  Or would this replace using split_at_mut when the sides cannot overlap?  If I'm reading it right, the one example in the docs doesn't need this method:
```rust
let (a, b) = bytes.split_at_mut(8);
b[..4].copy_from_slice(a[1..5]);
```

File: src/libcore/slice/mod.rs (hunk 1 above)
Comment (oconnor663) [reply src/libcore/slice/mod.rs]: I think without the casts this fails the borrow checker.

Comment (scottmcm) [reply src/libcore/slice/mod.rs]: Ah, yes, you're of course right.  I was thinking of [when they're not in variables](https://play.rust-lang.org/?gist=417fc09eb4dea89b518c1cc60f00adbf&version=stable&mode=debug&edition=2015):
```rust
        ptr::copy(
            slice.get_unchecked(src),
            slice.get_unchecked_mut(dest),
            count,
        );
```

**Comment by alexcrichton** [comment]
Looks good to me! I think there's some tidy/style issues though?

**Comment by oconnor663** [comment]
Line length should be fixed now.

**Comment by oconnor663** [comment]
Very silly copy-paste-o on my part. Should be fixed. I also decided I like @alexcrichton's name suggestion `copy_within` more, so I'm going with that for now.

**Comment by alexcrichton** [comment]
Thanks! Sorry I forgot to mention earlier, but could some semi-exhaustive tests be added for this as well? With `unsafe` in play it's always good to have a good suite of tests to handle issues!

**Comment by oconnor663** [comment]
I've added some tests, which are hopefully about to pass Travis. I could add more?

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Pull request merged**